compiler/qsc_eval/src/backend/noise_tests.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::{
+    backend::{Backend, SparseSim},
+    noise::PauliNoise,
+    state::{fmt_complex, format_state_id},
+};
+use expect_test::expect;
+use num_bigint::BigUint;
+use num_complex::Complex;
+use std::fmt::Write;
+
+#[test]
+fn pauli_noise() {
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0);
+    assert!(
+        noise
+            .expect("Expected construction of noiseless noise.")
+            .is_noiseless(),
+        "Expected noiseless noise."
+    );
+    let noise = PauliNoise::from_probabilities(1e-5, 0.0, 0.0);
+    assert!(
+        !noise
+            .expect("Expected construction of 1e-5 bit flip noise.")
+            .is_noiseless(),
+        "Expected noise to be noisy."
+    );
+    let noise = PauliNoise::from_probabilities(1.0, 0.0, 0.0);
+    assert!(
+        !noise
+            .expect("Expected construction of 1.0 bit flip noise.")
+            .is_noiseless(),
+        "Expected noise to be noisy."
+    );
+    let noise = PauliNoise::from_probabilities(0.01, 0.01, 0.01)
+        .expect("Expected construction of 0.01 depolarizing noise.");
+    assert!(!noise.is_noiseless(), "Expected noise to be noisy.");
+    assert!(
+        0.0 <= noise.distribution[0]
+            && noise.distribution[0] <= noise.distribution[1]
+            && noise.distribution[1] <= noise.distribution[2]
+            && noise.distribution[2] <= 1.1,
+        "Expected non-decreasing noise distribution."
+    );
+    let _ = PauliNoise::from_probabilities(-1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities -1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(1.0 + -1e-10, 0.1, 0.1)
+        .expect_err("Expected error for probabilities 1.0+1e-10, 0.1, 0.1.");
+    let _ = PauliNoise::from_probabilities(0.3, 0.4, 0.5)
+        .expect_err("Expected error for probabilities 0.3, 0.4, 0.5.");
+}
+
+#[test]
+fn noisy_simulator() {
+    let sim = SparseSim::new();
+    assert!(sim.is_noiseless(), "Expected noiseless simulator.");
+
+    let noise = PauliNoise::from_probabilities(0.0, 0.0, 0.0).expect("Cannot construct noise.");
```

Comment:
nitpick: these expect messages tend to follow the "expect as error message" style whereas I tend to agree with the leanings of the style recommendation in the Rust stdlib (see [Common Message Styles](https://doc.rust-lang.org/std/error/index.html#common-message-styles)) that "expect as precondition" is preferable. I've gotten used to reading the expect as almost like a sentence: I expect that <msg>. So this would be `.expect("noiseless should be constructable")` with starting with lower case. Since these are tests it's not critical, but especially if a test fails unpredictably the log output from something like CI might be our only clue, so having more natural langauge expectation could be useful. Food for thought, up to you if you fix these particular ones.

Reply:
I keep forgetting what the recommendation is. I reworked the errors.